src/core/src/GslBlockMatrix.C

```diff
@@ -0,0 +1,122 @@
+//-----------------------------------------------------------------------bl-
+//--------------------------------------------------------------------------
+//
+// QUESO - a library to support the Quantification of Uncertainty
+// for Estimation, Simulation and Optimization
+//
+// Copyright (C) 2008-2015 The PECOS Development Team
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the Version 2.1 GNU Lesser General
+// Public License as published by the Free Software Foundation.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc. 51 Franklin Street, Fifth Floor,
+// Boston, MA  02110-1301  USA
+//
+//-----------------------------------------------------------------------el-
+
+#include <queso/GslBlockMatrix.h>
+
+namespace QUESO {
+
+GslBlockMatrix::GslBlockMatrix(const FullEnvironment & env,
+    const std::vector<unsigned int> & blockSizes, double diagValue)
+  : m_env(env),
+    m_vectorSpaces(blockSizes.size()),
+    m_blocks(blockSizes.size())
+{
+  for (unsigned int i = 0; i < this->m_vectorSpaces.size(); i++) {
+    this->m_vectorSpaces[i] = new VectorSpace<GslVector, GslMatrix>(m_env,
+        "block_param_", blockSizes[i], NULL);
+    this->m_blocks[i] = new GslMatrix(this->m_vectorSpaces[i]->zeroVector(),
+        diagValue);
+  }
+}
+
+GslBlockMatrix::~GslBlockMatrix()
+{
+  for (unsigned int i = 0; i < this->m_vectorSpaces.size(); i++) {
+    delete this->m_blocks[i];
+    delete this->m_vectorSpaces[i];
+  }
+}
+
+GslMatrix &
+GslBlockMatrix::getBlock(unsigned int i) const
+{
+  return *(this->m_blocks[i]);
+}
+
+unsigned int
+GslBlockMatrix::numBlocks() const
+{
+  return this->m_blocks.size();
+}
+
+void
+GslBlockMatrix::invertMultiply(const GslVector & b, GslVector & x) const
+{
+  unsigned int totalCols = 0;
+
+  for (unsigned int i = 0; i < this->m_blocks.size(); i++) {
+    totalCols += this->m_blocks[i]->numCols();
+  }
+
+  if (totalCols != b.sizeLocal()) {
+    queso_error_msg("block matrix and rhs have incompatible sizes");
+  }
+
+  if (x.sizeLocal() != b.sizeLocal()) {
+    queso_error_msg("solution and rhs have incompatible sizes");
+  }
+
+  unsigned int blockOffset = 0;
+
+  // Do an invertMultiply for each block
+  for (unsigned int i = 0; i < this->m_blocks.size(); i++) {
+    GslVector blockRHS(this->m_vectorSpaces[i]->zeroVector());
```

Comment:
If this pops up in profiling, it might be worth implementing a version of invertMultiply that populates a solution vector that you give to it so we don't have to copy back and forth.